### PR TITLE
Improve precomputed hash optimizer

### PR DESCRIPTION
**Pull request by dain**
Optimizers can change the plan structure, adding or removing the need for
precomputed hashes, so only add precomputed hashes after plan is stable.

Remove unnecessary hash symbols before a hash build and drop them at
exchanges.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizersFactory.java [hunk 1]
```diff
@@ -107,6 +106,11 @@ public PlanOptimizersFactory(Metadata metadata, SqlParser sqlParser, FeaturesCon
         builder.add(new PruneUnreferencedOutputs());
         builder.add(new PruneIdentityProjections());
 
+        // NO NOT add optimizers that change the plan shape (computations) after this point
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizersFactory.java]: typo: ``DO NOT add..``

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java [hunk 2]
```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
...
-        return new ProjectNode(idAllocator.getNextId(), source, outputSymbols.build());
+        public Symbol getRequiredHashSymbol(List<Symbol> symbols)
+        {
+            Symbol hashSymbol = hashSymbols.get(symbols);
+            requireNonNull(hashSymbol, () -> "No hash symbol generated for " + symbols);
+            return hashSymbol;
+        }
+    }
+
+    private static Map<Symbol, Symbol> computeIdentityTranslations(Map<Symbol, Expression> assignments)
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java]: This code seems to be replicated in a bunch of places (`AddExchanges`, `PropertyDerivations`). Can we reuse this?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java [hunk 3]
```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
...
+                }
+            }
+
+            return new PlanWithProperties(new ProjectNode(idAllocator.getNextId(), source, assignments.build()), outputHashSymbols);
+        }
+
+        private PlanWithProperties plan(PlanNode node, Context context)
+        {
+            PlanWithProperties result = node.accept(this, context);
+            checkArgument(
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java]: use `checkState`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java [hunk 4]
```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
...
+            }
+
+            PlanNode source = result.getNode();
+            ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
+
+            Map<List<Symbol>, Symbol> outputHashSymbols = new HashMap<>();
+
+            // copy through all symbols from child, except for hash symbols not needed by the parent
+            for (Symbol symbol : source.getOutputSymbols()) {
+                List<Symbol> partitionSymbols = result.getHashSymbols().inverse().get(symbol);
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java]: extract `result.getHashSymbols().inverse()` out of the loop

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java [hunk 5]
```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
...
+                            node.getRowNumberSymbol(),
+                            node.getMaxRowCountPerPartition(),
+                            node.isPartial(),
+                            Optional.of(hashSymbol)),
+                    child.getHashSymbols());
         }
 
         @Override
-        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        public PlanWithProperties visitJoin(JoinNode node, Context context)
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java]: @erichwang  can you take a look at this method?

**Comment by nileema** [comment]
I've have @erichwang  look at the join stuff.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java [hunk 6]
```diff
@@ -69,13 +89,84 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
...
         requireNonNull(idAllocator, "idAllocator is null");
         if (SystemSessionProperties.isOptimizeHashGenerationEnabled(session)) {
-            return SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator, symbolAllocator, types), plan, null);
+            PlanWithProperties result = plan.accept(new Rewriter(idAllocator, symbolAllocator, types), new Context());
+            return result.getNode();
         }
         return plan;
     }
 
+    private static class Context
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java]: call this class something more meaningful

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java [hunk 7]
```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
...
+            Symbol rightHashSymbol = right.getRequiredHashSymbol(rightSymbols);
+
+            // build map of all hash symbols
+            Map<List<Symbol>, Symbol> allHashSymbols = new HashMap<>();
+            if (node.getType() == INNER || node.getType() == LEFT) {
+                allHashSymbols.putAll(left.getHashSymbols());
+            }
+            if (node.getType() == INNER || node.getType() == RIGHT) {
+                allHashSymbols.putAll(right.getHashSymbols());
+            }
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java]: leave a comment here about how FULL type is handled, and why we don't need to mention it

**Comment by nileema** [comment]
Looks good.

**Pull request closed without merging**